apis/v1alpha2/gateway_types.go

```diff
@@ -0,0 +1,835 @@
+	// The incoming hostname MUST be matched using the Hostname
+	// field for each Listener in order of most to least specific.
+	// That is, exact matches must be processed before wildcard
+	// matches.
+	//
+	// If this field specifies multiple Listeners that have the same
+	// Port value but are not compatible, the implementation must raise
+	// a "Conflicted" condition in the Listener status.
+	//
+	// Support: Core
+	//
+	// +listType=map
+	// +listMapKey=name
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=64
+	Listeners []Listener `json:"listeners"`
+
+	// Addresses requested for this gateway. This is optional and
+	// behavior can depend on the GatewayClass. If a value is set
+	// in the spec and the requested address is invalid, the
+	// GatewayClass MUST indicate this in the associated entry in
+	// GatewayStatus.Addresses.
+	//
+	// If no Addresses are specified, the GatewayClass may
+	// schedule the Gateway in an implementation-defined manner,
+	// assigning an appropriate set of Addresses.
+	//
+	// The GatewayClass MUST bind all Listeners to every
+	// GatewayAddress that it assigns to the Gateway.
+	//
+	// Support: Core
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	Addresses []GatewayAddress `json:"addresses,omitempty"`
+}
+
+// Listener embodies the concept of a logical endpoint where a Gateway can
+// accept network connections. Each listener in a Gateway must have a unique
+// combination of Hostname, Port, and Protocol. This will be enforced by a
+// validating webhook.
+type Listener struct {
+	// Name is the name of the Listener. If more than one Listener is present
+	// each Listener MUST specify a name. The names of Listeners MUST be unique
+	// within a Gateway.
+	//
+	// Support: Core
+	//
+	// +kubebuilder:validation:MaxLength=253
+	Name string `json:"name"`
+
+	// Hostname specifies the virtual hostname to match for protocol types that
+	// define this concept. When unspecified, "", or `*`, all hostnames are
+	// matched. This field can be omitted for protocols that don't require
+	// hostname based matching.
+	//
+	// For HTTPRoute objects, there is an interaction with the
+	// `spec.hostnames` array. When both listener and route specify hostnames,
+	// there must be an intersection between the values for a Route to be admitted.
+	// For example, a Gateway with `*.example.com` would admit a Route that included
+	// `foo.example.com` as a hostname, but not a Route that *only* included
+	// `foo.acme.io` as a hostname. A Route that included both `foo.example.com`
+	// and `foo.acme.io` would be admitted, but the `foo.acme.io` hostname would
+	// be silently ignored.
+	//
+	// Support: Core
+	//
+	// +optional
+	Hostname *Hostname `json:"hostname,omitempty"`
+
+	// Port is the network port. Multiple listeners may use the
+	// same port, subject to the Listener compatibility rules.
+	//
+	// Support: Core
+	Port PortNumber `json:"port"`
```

Comment:
hmm that (along with .HostName) limits the entire stack to L7 style proto. Is this a non-goal? 

What if i want a gateway for L4 load balancing?

Reply:
To clarify, are you suggesting that we should have the ability to listen on all ports?

Reply:
I am asking if the scope of the api is L7 protocols only?

Let us say i have one of those industry standards protos. But they are not as common as HTTP. maybe MQTT or similar. I want to build an ingress/reverse proxy for. Or to put more simply. I want to write some sort of exotic packet processing and route based on some logic. will i be able to use this apis?

Reply:
I think we're hoping that yes, this should be extendable for other APIs, but we are focussing right now on L7 HTTP use cases, with some thought put to L4 TCP and UDP use cases, along with the does-not-fit-cleanly-into-OSI-layer TLS use cases.

I'm in the process of doing an update to the concepts doc that may make this clearer, here is the relevant bit that's going to end up in there:
* HTTPRoute is for multiplexing HTTP or terminated HTTPS connections onto a single port or set of ports. It's intended for use in cases where you want to inspect the HTTP stream and use HTTP-level data for either routing or modification. (An example here is using HTTP Headers for routing, or modifying them in-flight).
* TLSRoute is for multiplexing TLS connections, discriminated via SNI, onto a single port or set or ports. It's intended for where you want to use the SNI as the main routing method, and are not interested in properties of the underlying connection (such as whether or not it is HTTP).
* TCPRoute (and UDPRoute) are intended for use as a mapping between a single port or set of ports and a single backend. In this case, there is no discriminator you can use to choose different backends on the same port, so each TCPRoute really needs a different port on the listener (in general, anyway).

This is what my thinking is here, anyway. In effect, in the order listed, the objects are different in terms of how deeply into the OSI layers of the connection they look for routing information. TCPRoute and UDPRoute consider only basic OSI layer 4 information, TLSRoute uses properties of the TLS connection only, and HTTPRoute uses all the panoply of properties that HTTP has available. 

While I can totally see that a Gateway could be used to describe an IPRoute, GRETunnelRoute, or something like that, we have put zero thought into the design for those things so far.